### PR TITLE
NNX migration prep (2/N): NNX utils and sharding utilities

### PR DESCRIPTION
**Pull request by ecnal-cienet**
# NNX Migration Route Map
1. ✅ Add NNX scaffolding: `pure_nnx` flag, `init_state_fn`, `TrainStateNNX`, NNX utils. Linen workflow unchanged. ([PR #3427](https://github.com/AI-Hypercomputer/maxtext/pull/3427))
2. ✅ **[This PR]** NNX sharding utilities: `get_abstract_state_nnx`, `get_named_sharding_nnx`, `set_named_sharding_nnx`, `get_partition_spec_nnx`, `get_mesh_from_config`. ([PR #3470](https://github.com/AI-Hypercomputer/maxtext/pull/3470))
3. ✅ NNX fully supported end-to-end: `TrainStateNNX`, model creation, gradient accumulation, checkpointing, and training loop dispatch. ([PR #3500](https://github.com/AI-Hypercomputer/maxtext/pull/3500))
4. ✅ NNX sharding diagnostics and bidirectional Linen↔NNX checkpoint conversion utilities. ([PR #3525](https://github.com/AI-Hypercomputer/maxtext/pull/3525))
5. ❌ NNX post-training fixes: MultimodalInput unpacking, scalar LR guard, nested NNX transform workaround.
6. ❌ Enable NNX by default; fix unit and integration test failures.
7. ❌ Remove Linen-specific code paths and NNX compatibility flags.

# Description
> **Note:** This is the second in a series of NNX migration PRs. Pure NNX training is **not yet implemented** — all NNX code paths currently raise `NotImplementedError`. This PR only introduces the structural scaffolding needed for subsequent patches to plug in NNX logic without modifying shared infrastructure.

- **NNX sharding utilities** (`maxtext_utils_nnx.py`) — Functions to manipulate NNX model shardings using abstract model state: `get_named_sharding_nnx`, `set_named_sharding_nnx`, `get_partition_spec_nnx`, and memory movement helpers (`move_memory_to_host` / `move_memory_to_device`).
- **`get_abstract_state` NNX path** — Added `get_abstract_state_nnx` to `maxtext_utils.py`, which uses `nnx.get_abstract_model` to return a flat `nnx.State` (rather than a full `TrainStateNNX`), and updated `get_abstract_state` to dispatch to it when `pure_nnx=True`.
- **`maxtext_utils.get_mesh_from_config()`** — Extracted mesh creation into a standalone function with unit tests.
- **Unit tests** — Added `tests/unit/maxtext_utils_nnx_test.py` and extended `tests/unit/maxtext_utils_test.py` to cover the new mesh and sharding utilities.

**Note on Flax deprecation warnings:**                                                                                                                          
  Flax v0.12 emits `DeprecationWarning` for `.value` access and `VariableState`. These are intentionally left unaddressed because post-training currently requires Flax v0.11 compatibility.

# Tests
```
pytest tests/unit/maxtext_utils_nnx_test.py tests/unit/maxtext_utils_test.py -v 
pytest tests/unit/maxtext_utils_nnx_test.py tests/unit/maxtext_utils_nnx_test.py -v 
```

# Pre-train Test Result
[View Result](https://ecnal-cienet.github.io/maxview/feat_migrate_nnx_utils_20260415_230805/index.html)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/model_creation_utils.py [hunk 1]
```diff
@@ -260,16 +251,10 @@ def _create_model(rng_key=None):
...
   def _create_model(mesh: Mesh | None = None, model_mode: str = MODEL_MODE_TRAIN, rng_key: jax.Array | None = None):
-    if rng_key is None:
-      rng_key = jax.random.PRNGKey(config.init_weights_seed)
-
-    if model_mode == MODEL_MODE_TRAIN:
-      rngs = nnx.Rngs(params=rng_key, dropout=1)
-    else:
-      rngs = nnx.Rngs(params=rng_key)  # disable dropout RNG for inference
-
+    rngs = maxtext_utils_nnx.create_nnx_rngs(config, is_training=is_training, rng_key=rng_key)
```
Comment (bvandermoon) [inline src/maxtext/utils/model_creation_utils.py]: Why are the training rngs different than the removed snippet? https://github.com/AI-Hypercomputer/maxtext/blob/992b4e193e86a144f389077a90cdbba24603f4e2/src/maxtext/utils/maxtext_utils_nnx.py#L46

Also, can you change `create_nnx_rngs` to accept `model_mode` directly? That way we wouldn't need to create a second variable that gives almost the same info

Comment (ecnal-cienet) [reply src/maxtext/utils/model_creation_utils.py]: Hi Branden, 
1. The old code `nnx.Rngs(params=key, dropout=1, aqt=2)` passed integer literals 1 and 2 as the dropout and aqt RNG streams. nnx.Rngs accepts these as-is and folds them into the key internally — but it silently collapses both streams to near-identical keys (fold_in(key, 1) and fold_in(key, 2) are very close in key-space when the base key varies only at the low bits). More importantly, those integers aren't JAX arrays, so they bypass JAX's PRNG splitting guarantees.
The new code uses jax.random.fold_in(rng_key, 0/1/2) explicitly, which is the correct way to derive independent, statistically-uncorrelated keys for each stream from a single seed. Also added aqt as a third stream (needed for quantization-aware training). Inference mode now returns only nnx.Rngs(params=rng_key) — no dropout or aqt, since those aren't used during inference.

2. Done — changed create_nnx_rngs to accept model_mode: str (defaulting to MODEL_MODE_TRAIN) instead of is_training: bool. Updated all call sites in model_creation_utils.py and maxtext_utils_nnx_test.py accordingly.

File: src/maxtext/utils/maxtext_utils_nnx.py [hunk 2]
```diff
@@ -22,17 +22,18 @@
...
-    config: pyconfig.HyperParameters, is_training: bool = True, rng_key: jax.Array | None = None
+    config: pyconfig.HyperParameters, model_mode: str = MODEL_MODE_TRAIN, rng_key: jax.Array | None = None
 ) -> nnx.Rngs:
   """
   Create NNX Rngs
 
   Args:
     config: the configuration
-    is_training: if the Rngs are for training
+    model_mode: the model mode (train or inference)
```
Comment (bvandermoon) [inline src/maxtext/utils/maxtext_utils_nnx.py]: Nit: Could you remove the `(train or inference)` piece here? Since it is actually `MODEL_MODE_TRAIN`, `MODEL_MODE_PREFILL`, etc. You could point to https://github.com/AI-Hypercomputer/maxtext/blob/9c68b1abd479c4e95fc6e66f0c7a4235a9507a9e/src/maxtext/common/common_types.py if you want. I am ok to just keep this as `model_mode: the model mode` though since we use this throughout MaxText

Comment (ecnal-cienet) [reply src/maxtext/utils/maxtext_utils_nnx.py]: Done. Thanks Branden.
